### PR TITLE
Replace VUID error text map with struct to improve build times

### PR DESCRIPTION
**Pull request by daveh-lunarg**
Replaces the initialized-in-header `std::unordered_map` container holding VUID->spec_text mapping with a simple string array. VUID and spec text strings are now imported as vanilla `char*` string literals. Also cleaned up unnecessary includes of the error text header, previously needed for enum visibility but now obsoleted.

This change is a workaround for slow build times observed with use of the header-resident map, in particular on 32-bit Android builds. It avoids the compile-time problem, and improves run-time performance on validation tests as well.

### CODE REVIEW

File: layers/vk_layer_logging.h [hunk 1]
```diff
@@ -896,6 +898,14 @@ static inline bool log_msg(const debug_report_data *debug_data, VkFlags msg_flag
 
     std::string str_plus_spec_text(str ? str : "Allocation failure");
 
+    // first-time-through initialization of error text map
```
Comment (jzulauf-lunarg) [inline layers/vk_layer_logging.h]: need a or mutex/atomic CAS/ or call_once to guard initialization. Easiest is to put initialization in a function and call it with std::call_once https://en.cppreference.com/w/cpp/thread/call_once

Comment (jzulauf-lunarg) [reply layers/vk_layer_logging.h]: More background/usage http://www.modernescpp.com/index.php/thread-safe-initialization-of-data

Comment (daveh-lunarg) [reply layers/vk_layer_logging.h]: Added call_once.

File: layers/vk_layer_logging.h [hunk 2]
```diff
@@ -877,6 +887,9 @@ static inline bool log_msg(const debug_report_data *debug_data, VkFlags msg_flag
 #endif
 static inline bool log_msg(const debug_report_data *debug_data, VkFlags msg_flags, VkDebugReportObjectTypeEXT object_type,
                            uint64_t src_object, std::string vuid_text, const char *format, ...) {
+    static std::unordered_map<std::string, std::string> vuid_to_error_text_map;
+    std::once_flag map_flag;
```
Comment (jzulauf-lunarg) [inline layers/vk_layer_logging.h]: needs to be static here, or at compilation unit scope, as this is the persistent marker of "doneness"

**Comment by karl-lunarg** [comment]
Looks good, pending the fixes.  I like the faster Travis-CI times.

File: layers/vk_validation_error_messages.h [hunk 3]
```diff
@@ -795,9 +796,9 @@ std::unordered_map<std::string, std::string> vuid_to_error_text_map {
...
     {"VUID-VkGraphicsPipelineCreateInfo-pDynamicStates-00755[(VK_EXT_depth_range_unrestricted)]", "If the VK_EXT_depth_range_unrestricted extension is not enabled and no element of the pDynamicStates member of pDynamicState is VK_DYNAMIC_STATE_DEPTH_BOUNDS, and the depthBoundsTestEnable member of pDepthStencilState is VK_TRUE, the minDepthBounds and maxDepthBounds members of pDepthStencilState must be between 0.0 and 1.0, inclusive (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-pDynamicStates-00755)"},
+    {"VUID-VkGraphicsPipelineCreateInfo-pDynamicStates-00755[!(VK_EXT_depth_range_unrestricted)]", "If no element of the pDynamicStates member of pDynamicState is VK_DYNAMIC_STATE_DEPTH_BOUNDS, and the depthBoundsTestEnable member of pDepthStencilState is VK_TRUE, the minDepthBounds and maxDepthBounds members of pDepthStencilState must be between 0.0 and 1.0, inclusive (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-pDynamicStates-00755)"},
```
Comment (jzulauf-lunarg) [inline layers/vk_validation_error_messages.h]: Not that it has anything to do with this fix, how (is?) a VUID of this form used?

Comment (daveh-lunarg) [reply layers/vk_validation_error_messages.h]: I think that's a JSON error, or a failure of the parsing in the validation_stats script. I noticed in in the diff but haven't triaged it yet.  The diff I think is just the order in which the 755 text was exported, but there's an underlying bug to fix.

**Comment by daveh-lunarg** [comment]
I missed the static decoration, but it looks like the Linux builds are just unhappy with call_once in general. It also seems to me that call_once protects against multiple initializations, but I don't think it helps to avoid a second thread seeing the partially-initialized map? That requires some kind of mutex around the .empty() call - unpopular I'm sure, but again it's in the exception path.

TBH, I'm leaning now toward just removing the deferred-initialization in favor of one of the static initialization variants that Jeremy and Cody tried that don't trigger the build problem.  If there's is one that all the compiler flavors will agree on, at least.  Discuss tomorrow.

**Comment by jzulauf-lunarg** [comment]
@daveh-lunarg -- call once isn't supported by VS2013 either.  The other choice is a static mutex locked before the "empty" check, you could short-circuit the lock with a static bool "initialized" set by the initializing thread.
```
static mutex map_mutex;  // Statics are guaranteed to be initialized once
static bool map_initialized = false;
if ( !map_initialized) {
   lock_guard(map_mutex);
   if (!map.empty) {
    init_map(map);
    map_initialized = true;
  }
}
```

Which is what once_flag and call once have to be doing under the covers essentially.

File: layers/vk_layer_logging.h [hunk 4]
```diff
@@ -868,6 +871,13 @@ static inline int vasprintf(char **strp, char const *fmt, va_list ap) {
 }
 #endif
 
+static inline void InitErrorTextMap(std::unordered_map<std::string, std::string> &error_map) {
```
Comment (jzulauf-lunarg) [inline layers/vk_layer_logging.h]: I wonder if flagging the function inline is confusing the Linux compiler that the function isn't a Callable (though that would seem like a bug). 

In any case VS 2013 won't support call_once -- sorry for the bum steer.

File: layers/vk_layer_logging.h [hunk 5]
```diff
@@ -896,6 +909,11 @@ static inline bool log_msg(const debug_report_data *debug_data, VkFlags msg_flag
 
     std::string str_plus_spec_text(str ? str : "Allocation failure");
 
+    // first-time-through initialization of error text map
+    if (vuid_to_error_text_map.empty()) {
```
Comment (jzulauf-lunarg) [inline layers/vk_layer_logging.h]: For future reference the .empty() test is not required.  call_once either

* calls the function (if no thread has)
* blocks until the caller finishes and returns (modulo exceptions)
* returns if the caller has completed

So the test is redundant -- if of course we could use call_once...  (which starting with VS 2015 we can)

**Comment by daveh-lunarg** [comment]
Rather than try to fix the synchronized run-time initialization in a manner that all our compiler flavors can tolerate, I've just switched to a lowest-common-denominator static initialization with a linear search at error reporting time. This turns out to be win/win for both avoiding the compile-time bug, and improving run-time performance of the layer validation tests. Wall-clock time for `layer_validation_tests` decrease of about 45% vs master (Windows, debug) with this change.

Apparently the linear search time penalty is dwarfed by the perf savings from avoiding the map build altogether, and the frequent setup/teardown of the lvt is well aligned to demonstrate the effect.

**Comment by cnorthrop** [comment]
Tested locally, seeing both compile and runtime perf benefit, thanks Dave.

File: layers/vk_validation_error_messages.h (hunk 3 above)
Comment (daveh-lunarg) [reply layers/vk_validation_error_messages.h]: @jzulauf-lunarg : Not a bug, it's a feature! I hate it when I don't recognize something I wrote just a few months before...

There are about a half dozen VUIDs with multiple entries in the validusage.json file, typically in 2 flavors with/without a particular extension.  Any VUID that has multiple definitions goes into the header as a generic version, plus one version with extension critera appended per definition. The per-extension versions are not considered by log_msg, at the moment.

**Pull request merged**